pkg/icingadb/runtime_updates.go

```diff
@@ -70,22 +73,114 @@ func (r *RuntimeUpdates) Sync(ctx context.Context, factoryFuncs []contracts.Enti
 		updateMessages := make(chan redis.XMessage, bulkSize)
 		upsertEntities := make(chan contracts.Entity, bulkSize)
 		deleteIds := make(chan interface{}, bulkSize)
+		threshold := com.CommonBulkThreshold
+		var upserted chan contracts.Entity
+		var upsertedFifo chan contracts.Entity
+		var upsertedLog chan contracts.Entity
+		var deleted chan interface{}
+		var deletedFifo chan interface{}
+		var deletedLog chan interface{}
+
+		if !allowParallel {
+			threshold = -1
+			upserted = make(chan contracts.Entity, 1)
+			upsertedFifo = make(chan contracts.Entity, 1)
+			upsertedLog = make(chan contracts.Entity, 1)
+			deleted = make(chan interface{}, 1)
+			deletedFifo = make(chan interface{}, 1)
+			deletedLog = make(chan interface{}, 1)
+
+			g.Go(func() error {
+				for {
+					select {
+					case v, ok := <-upserted:
+						if !ok {
+							return nil
+						}
+
+						select {
+						case upsertedLog <- v:
+						case <-ctx.Done():
+							return ctx.Err()
+						}
+
+						select {
+						case upsertedFifo <- v:
+						case <-ctx.Done():
+							return ctx.Err()
+						}
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+				}
+			})
+
+			g.Go(func() error {
+				for {
+					select {
+					case v, ok := <-deleted:
+						if !ok {
+							return nil
+						}
+
+						select {
+						case deletedLog <- v:
+						case <-ctx.Done():
+							return ctx.Err()
+						}
+
+						select {
+						case deletedFifo <- v:
+						case <-ctx.Done():
+							return ctx.Err()
+						}
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+				}
+			})
+		} else {
+			upserted = make(chan contracts.Entity)
+			deleted = make(chan interface{})
+			upsertedLog = upserted
+			deletedLog = deleted
```

Comment:
And if `allowParallel`, then `upsertedLog = upserted`, so both might observe that the channel is not yet closed and attempt to close it resulting in a panic.